qcodes_contrib_drivers/drivers/CopperMountain/M5180.py

```diff
@@ -0,0 +1,564 @@
+# This Python file uses the following encoding: utf-8
+# Etienne Dumur <etienne.dumur@gmail.com>, august 2020
+# Simon Zihlmannr <zihlmann.simon@gmail.com>, february/march 2021
+import logging
+import numpy as np
+from typing import Tuple
+
+from qcodes import VisaInstrument
+from qcodes.utils.validators import Numbers, Enum, Ints
+from qcodes.utils.helpers import create_on_off_val_mapping
+
+from qcodes.instrument.parameter import (
+    MultiParameter,
+    ParamRawDataType
+)
+
+log = logging.getLogger(__name__)
+
+class FrequencySweepMagPhase(MultiParameter):
+    """
+    Sweep that returns magnitude and phase.
+    """
+
+    def __init__(self,
+        name: str,
+        start: float,
+        stop: float,
+        npts: int,
+        instrument: Optional[InstrumentBase] = "M5180",
```

Comment:
How do I have to declare the instrument class here? I guess not as a string. Is `instrument: Optional[InstrumentBase] = M5180,` fine?

Unfortunately, I have not found how to do it, neither in the example notebook nor in a similar driver in qcodes (ZNB driver from R&S).

Reply:
@FarBo could you maybe enlighten on how to solve this issue? Thanks in advance.

Reply:
Sure,

Do you really expect that another instrument other than M5180 uses this class? 
Also, this class uses instrument argument inside, so instrument cannot be Optional.

you probably can do such thing:

from qcodes.instrument import InstrumentBase
...
class PointMagPhase(MultiParameter):
...
instrument: InstrumentBase = "M5180"

Reply:
Not really since it has some instrument specific commands. So I guess it would be better to hard code it. I'll try your suggestion. Thanks.

Reply:
Sorry, there was a little bit of mislead in my first suggestion

You can not use the instrument M5180 here because it needs to be an instance of  M5180 class, which means that there should be a name and address passed to it, and it is not possible in the driver level.

Actually, here you need to say:
**instrument: Optional['M5180'] = None** (please remember to import Optional from typing)
Then inside the init you should have:
**assert instrument is not None** 

The default value to instrument can be only None in this structure.

Since having this default is not going to serve what you want (as explained above), you can simply say:
**instrument: 'M5180'**

Reply:
Hi @FarBo,
Thanks for your additional inputs. Though I am not sure if I fully understood your comment correctly. Here is what I have so far, but I still get some errors with mypy.
Code:
```
class FrequencySweepMagPhase(MultiParameter):
    """
    Sweep that returns magnitude and phase.
    """

    def __init__(self,
        name: str,
        start: float,
        stop: float,
        npts: int,
        instrument: "M5180",
        ) -> None:
        """
        Linear frequency sweep that returns magnitude and phase for a single
        trace.

        Args:
            name (str): Name of the linear frequency sweep
            start (float): Start frequency of linear sweep
            stop (float): Stop frequency of linear sweep
            npts (float): Number of points of linear sweep
            instrument: Instrument to which sweep is bound to.
        """
        super().__init__(
            name,
            instrument=instrument,
            names=(
                f"{instrument.short_name}_{name}_magnitude",
                f"{instrument.short_name}_{name}_phase"),
            labels=(
                f"{instrument.short_name} {name} magnitude",
                f"{instrument.short_name} {name} phase",
            ),
            units=("dB", "rad"),
            setpoint_units=(("Hz",), ("Hz",)),
            setpoint_labels=(
                (f"{instrument.short_name} frequency",),
                (f"{instrument.short_name} frequency",),
            ),
            setpoint_names=(
                (f"{instrument.short_name}_frequency",),
                (f"{instrument.short_name}_frequency",),
            ),
            shapes=((npts,), (npts,),),
        )
        self.set_sweep(start, stop, npts)

    def set_sweep(self, start: float, stop: float, npts: int) -> None:
        """Updates the setpoints and shapes based on start, stop and npts.

        Args:
            start (float): start frequency
            stop (float): stop frequency
            npts (int): number of points
        """
        f = tuple(np.linspace(int(start), int(stop), num=npts))
        self.setpoints = ((f,), (f,))
        self.shapes = ((npts,), (npts,))

    def get_raw(self) -> Tuple[ParamRawDataType, ParamRawDataType]:
        """Gets data from instrument

        Returns:
            Tuple[ParamRawDataType, ...]: magnitude, phase
        """
        self.instrument.write('CALC1:PAR:COUN 1') # 1 trace
        self.instrument.write('CALC1:PAR1:DEF {}'.format(self.name))
        self.instrument.write('CALC1:TRAC1:FORM SMITH')  # Trace format
        self.instrument.trigger_source('bus') # set the trigger to bus
        self.instrument.write('TRIG:SEQ:SING') # Trigger a single sweep
        self.instrument.ask('*OPC?') # Wait for measurement to complete

        # get data from instrument
        sxx = self.instrument.ask("CALC1:TRAC1:DATA:FDAT?")

        # Get data as numpy array
        sxx = np.fromstring(sxx, dtype=float, sep=',')
        sxx = sxx[0::2] + 1j*sxx[1::2]

        return self.instrument._db(sxx), np.unwrap(np.angle(sxx))
```

Error (they all appear in the get_raw method on the lines with `self.instrument`):
```
M5180.py:85: error: Item "None" of "Optional[InstrumentBase]" has no attribute "write"
M5180.py:86: error: Item "None" of "Optional[InstrumentBase]" has no attribute "write"
M5180.py:87: error: Item "None" of "Optional[InstrumentBase]" has no attribute "write"
M5180.py:88: error: Item "None" of "Optional[InstrumentBase]" has no attribute "trigger_source"
M5180.py:89: error: Item "None" of "Optional[InstrumentBase]" has no attribute "write"
M5180.py:90: error: Item "None" of "Optional[InstrumentBase]" has no attribute "ask"
M5180.py:93: error: Item "None" of "Optional[InstrumentBase]" has no attribute "ask"
M5180.py:99: error: Item "None" of "Optional[InstrumentBase]" has no attribute "_db"
```

Could you maybe help me to fix this?

Reply:
Because of the way that that inheritance works mypy cannot see that in you case `self.instrument` is an instance of the M5180 but only that it (like in all MultiParameters) is either an instance of `InstrumentBase` or None. The easiest way to fix this is probably to help mypy a bit by asserting that this is really an `M5180`

I think that can be done by adding the following line to the start of your get_raw method

```
assert isinstance(self.instrument, M5180)
```

Reply:
@jenshnielsen thanks a lot for the explanation!